src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs

```diff
@@ -1253,6 +1273,59 @@ private ICollectionExpressionOperation CreateBoundCollectionExpression(BoundColl
                 }
             }
+
+            static ImmutableArray<IOperation> getConstructArguments(
+                CSharpOperationFactory @this, BoundCollectionExpression expr)
+            {
+                var collectionCreation = expr.CollectionCreation;
+                while (collectionCreation is BoundConversion conversion)
+                    collectionCreation = conversion.Operand;
+
+                if (collectionCreation is BoundObjectCreationExpression objectCreation)
+                {
+                    // Match the logic in CreateBoundObjectCreationOperation which does not DeriveArguments in the case of an
+                    // problems encountered in binding.
+                    Debug.Assert(!objectCreation.Type.IsAnonymousType);
+                    if (!CanDeriveObjectCreationExpressionArguments(objectCreation))
+                        return @this.CreateFromArray<BoundNode, IOperation>(((IBoundInvalidNode)objectCreation).InvalidNodeChildren);
+
+                    return ImmutableArray<IOperation>.CastUp(@this.DeriveArguments(collectionCreation));
+                }
+
+                if (collectionCreation is BoundCall boundCall)
+                {
+                    // Match the logic in CreateBoundCallOperation which does not DeriveArguments in the case of an
+                    // erroneous call node.
+                    if (boundCall.IsErroneousNode)
+                    {
+                        var array = @this.CreateFromArray<BoundNode, IOperation>(((IBoundInvalidNode)boundCall).InvalidNodeChildren);
+                        return array.WhereAsArray(o => o is not IPlaceholderOperation);
+                    }
+
+                    var arguments = @this.DeriveArguments(collectionCreation);
+
+                    // With a CollectionBuilder, the last argument will be a placeholder where the .Elements will go.
+                    // We do *not* want to include that information in the Arguments we return.
+                    Debug.Assert(arguments is [.., IArgumentOperation { Value: IPlaceholderOperation { PlaceholderKind: PlaceholderKind.Unspecified } }], "We should always have at least one argument (the placeholder elements).");
+                    var slicedArguments = arguments is [.. var normalArguments, _]
+                        ? normalArguments
+                        : arguments;
+
+                    return ImmutableArray<IOperation>.CastUp(slicedArguments);
+                }
+
+                if (collectionCreation is BoundNewT boundNewT)
+                    return [];
+
+                if (collectionCreation is BoundBadExpression boundBad)
+                    return @this.CreateFromArray<BoundExpression, IOperation>(boundBad.ChildBoundNodes);
+
+                if (collectionCreation is null)
+                    return [];
+
+                Debug.Fail("Unhandled case: " + collectionCreation.GetType());
```

Comment:
>Debug.Fail("Unhandled case: " + collectionCreation.GetType()); [](http://example.com/codeflow?start=16&length=62)

Please add `Validate` method to the ```BoundCollectionExpression``` node by adding `HasValidate="true"` attribute in `BoundNodes.xml`, and assert possible values that could be in ```CollectionCreation``` according to this method.

Reply:
Nice.  7bdad0d4abe